File: ejemplos-cpp/src/Ejemplo1.cpp
#include <opencv2/core.hpp>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <cstddef>
#include <iostream>
#include <string>
#include <vector>

#include "utils/util_opencv.hpp"
#include "utils/util_strings.hpp"
#include "utils/util_ui.hpp"

void procesar_imagen(const std::string &filename) {
	std::cout << "abriendo " << filename << std::endl;
	cv::Mat imagen_color = cv::imread(filename, cv::IMREAD_COLOR);
	if (imagen_color.empty()) {
		std::cout << "no puedo abrir imagen " << filename << std::endl;
		return;
	}
	cv::Mat imagen_gris;
	cv::cvtColor(imagen_color, imagen_gris, cv::COLOR_BGR2GRAY);
	cv::Mat imagen_bin;
	int threshold = cv::threshold(imagen_gris, imagen_bin, 0, 255,
			cv::THRESH_BINARY | cv::THRESH_OTSU);
	std::string window_name = basename(filename);
	cv::imshow(window_name, imagen_color);
	agregar_texto(imagen_bin, imagen_bin.cols / 2, 2,
			"OTSU=" + std::to_string(threshold));
	cv::imshow(window_name + " BINARIA", imagen_bin);
	std::cout << window_name << ": size=" << imagen_color.size()
			<< " threshold_otsu=" << threshold << std::endl;
}
void procesar_archivos(std::vector<std::string> filenames) {
	if (filenames.size() == 0)
		filenames = ui_select_filenames();
	if (filenames.size() == 0)
		return;
	for (std::string filename : filenames)
		procesar_imagen(filename);
	std::cout << "Presione una tecla..." << std::endl;
	cv::waitKey(0);
	cv::destroyAllWindows();
}

void procesar(const std::vector<std::string> &args) {
	std::cout << "CC5213 - Ejemplo 1 OTSU" << std::endl;
	std::cout << "Usando OpenCV " << CV_VERSION << std::endl;
	if (print_help(args)) {
		std::cout << "  Uso: " << basename(args[0]) << " [filenames]"
				<< std::endl;
	} else {
		std::vector<std::string> filenames(args.begin() + 1, args.end());
		procesar_archivos(filenames);
	}
	std::cout << "fin" << std::endl;
}

int main(int argc, char **argv) {
	try {
		std::vector<std::string> args = get_args_vector(argc, argv);
		procesar(args);
	} catch (const std::exception &ex) {
		std::cout << "Ha ocurrido un ERROR: " << ex.what() << std::endl;
	} catch (...) {
		std::cout << "Ha ocurrido ERROR desconocido" << std::endl;
	}
	return 0;
}

File: ejemplos-python/Ejemplo1.py
import sys
import cv2
import os
from utils import utils

def procesar_imagen(filename):
    print("abriendo {}".format(filename))
    imagen_color = cv2.imread(filename, cv2.IMREAD_COLOR)
    if imagen_color is None:
        print("no puedo abrir imagen {}".format(filename))
        return
    imagen_gris = cv2.cvtColor(imagen_color, cv2.COLOR_BGR2GRAY)
    threshold, imagen_bin = cv2.threshold(imagen_gris, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)
    window_name = os.path.basename(filename)
    cv2.imshow(window_name, imagen_color)
    cv2.imshow(window_name + " BINARIA", imagen_bin)
    print("{}: size={} threshold_otsu={}".format(window_name, imagen_color.shape, threshold))

def procesar_archivos(filenames):
    if len(filenames) == 0:
        filenames = utils.ui_select_filenames()
    if len(filenames) == 0:
        return
    for filename in filenames:
        procesar_imagen(filename)
    print("Presione una tecla...")
    cv2.waitKey(0)
    cv2.destroyAllWindows()


print("CC5213 - Ejemplo 1 OTSU")
print("Usando OpenCV {} con Python {}.{}.{}".format(cv2.__version__, sys.version_info.major, sys.version_info.minor, sys.version_info.micro))

if utils.print_help(sys.argv):
    print("Uso: {} [filenames]".format(sys.argv[0]))
else:
    procesar_archivos(sys.argv[1:])

print("fin")
